lib/core/src/Cardano/Wallet/Primitive/Slotting.hs

```diff
@@ -0,0 +1,262 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Contains tools for converting between @SlotNo@, @EpochNo@, @SlotInEpoch@,
+-- @UTCTime@.
+
+module Cardano.Wallet.Primitive.Slotting
+    ( -- * New api using ouroboros-concensus
+      epochOf
+    , singleEraInterpreter
+    , runQuery
+
+      -- * Old functions
+    , unsafeEpochNo
+    , epochStartTime
+    , epochPred
+    , epochSucc
+    , SlotParameters (..)
+    , slotParams
+    , flatSlot
+    , fromFlatSlot
+    , slotStartTime
+    , slotCeiling
+    , slotFloor
+    , slotAt
+    , slotDifference
+    , slotPred
+    , slotSucc
+    , slotMinBound
+    , slotRangeFromTimeRange
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( ActiveSlotCoefficient (..)
+    , EpochLength (..)
+    , EpochNo (..)
+    , GenesisParameters (..)
+    , Range (..)
+    , SlotId (..)
+    , SlotInEpoch (..)
+    , SlotLength (..)
+    , StartTime (..)
+    , unsafeEpochNo
+    )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Time.Clock
+    ( NominalDiffTime, UTCTime, addUTCTime, diffUTCTime )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+import Ouroboros.Consensus.HardFork.History.EraParams
+    ( EraParams (..), noLowerBoundSafeZone )
+import Ouroboros.Consensus.HardFork.History.Qry
+    ( Qry, runQuery, slotToEpoch )
+import Ouroboros.Consensus.HardFork.History.Summary
+    ( Summary (..), neverForksSummary )
```

Comment:
Importing of Ouroboros.Consensus doesn't seem right for the cardano-wallet-core package - it is cardano-node backend specific. Is there any way to move this to the cardano-wallet-shelley package?

Reply:
Core already depends on ouroboros-network: https://github.com/input-output-hk/cardano-wallet/blob/315b875cd3c6a78b8c4d4e58bfec6bd4d9acffcd/lib/core/src/Ouroboros/Network/Client/Wallet.hs#L12-L20


- To avoid having `Qry` in core, we'd need to duplicate the interface
- Jormungandr can satisfy the `Qry` interface

I do think it is fine.

_Slightly_ related thread: [point vs forcing jormungandr to use SlotNo](https://input-output-rnd.slack.com/archives/GBT05825V/p1594312980015000)

Reply:
The boundary between jormungandr and byron/shelley could also be revisited later.